### PR TITLE
[feat] 4/N Ragged inference - Pull in Triton v2 matmul and clean it up a bit

### PR DESCRIPTION
**Pull request by nottombrown**

### CODE REVIEW

File: tests/ragged_inference/test_ragged_qk_dotprod.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("dtype", [torch.float32])
+def test_mem_efficient_attention_parity(shape, dtype):
+    q = torch.rand(shape, dtype=dtype, device=torch.device("cuda"))
+    k = torch.rand(shape, dtype=dtype, device=torch.device("cuda"))
+
+    res_pytorch = attention_pytorch(q, k)
+    res_me = mem_efficient_fw(q=q, k=k)
+    # print(res_me)
+    assert torch.mean(torch.abs(res_pytorch - res_me)) < 0.2
```
Comment (blefaudeux) [inline tests/ragged_inference/test_ragged_qk_dotprod.py:67]: @nottombrown so this works not with a catch a little more stringent ? (toch.max()) torch.allclose() would probably be a good way to really test that

File: xformers/triton/ragged_inference/triton_v2_matmul.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        )
+    ]
+
+
+
+@triton.autotune(
+    # configs=get_all_configs(),
+    configs=get_fast_dev_configs(),
+    key=["M", "N", "K"],
+    prune_configs_by={
```
Comment (blefaudeux) [inline xformers/triton/ragged_inference/triton_v2_matmul.py:99]: oh, this is new !

File: xformers/triton/ragged_inference/triton_v2_qk_dotprod.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    acc_tile = tl.zeros((BLOCK_Q, BLOCK_K), dtype=tl.float32)
+    rd = tl.arange(0, BLOCK_D)
+
+    q_ptr_tile = q_ptr + (rq[:, None] * stride_ctx_q + rd[None, :] * stride_d)
+    k_ptr_tile = k_ptr + (rd[:, None] * stride_d + rk[None, :] * stride_ctx_k)
+
+    # acc_tile = tl.load(k_ptr_tile)
+
+    for d_max_offset in range(d_model, 0, -BLOCK_D):
+        q_tile = tl.load(q_ptr_tile, mask=rd[None, :] < d_max_offset, other=0.0)
```
Comment (blefaudeux) [inline xformers/triton/ragged_inference/triton_v2_qk_dotprod.py:156]: (masking) guess here is that BLOCK_Q, BLOCK_K will not tile beyond their respective dimensions (Q and K), which is completely possible, but FYI if you get a mem error on weird sizes that could be why, since this just guards on D. Just mentioning if this helps, not a worry

File: xformers/triton/ragged_inference/triton_v2_qk_dotprod.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    n_ctx_q, d_model = query.shape
+    n_ctx_k, d_model_k = key.shape
+    assert d_model == d_model_k, f"{query.shape=} {key.shape=}"
+
+    # allocates output
+    scores_out = torch.empty((n_ctx_q, n_ctx_k), device=device, dtype=query.dtype)
+
+    # Stride along the d_model dimension
+    stride_d = query.stride(1)
+    assert stride_d == key.stride(1), f"{stride_d=}, {key.stride(1)=}"
```
Comment (blefaudeux) [inline xformers/triton/ragged_inference/triton_v2_qk_dotprod.py:200]: stride_d here will be 1, no ? the tensors are made contiguous above

File: tests/ragged_inference/test_triton_v2_ragged_qk_dotprod.py [hunk 5]
```diff
@@ -109,9 +112,77 @@ def test_ragged_qk_dotprod_multiple_seqs():
...
+        n_ctx_k_per_seq=key.n_ctx_per_seq,
+    )
+
+    for _ in range(3):
+        out = ragged_qk_dotprod(query, key, lut)
+
+    torch.cuda.synchronize()
+    started_at = time.time()
+    for _ in range(n_iters):
+        out = ragged_qk_dotprod(query, key, lut)
```
Comment (blefaudeux) [inline tests/ragged_inference/test_triton_v2_ragged_qk_dotprod.py:156]: nit, but there could be caching/warm issues here (would not change the perf that much, but FYI). Triton exposes a great too for this, we use it in most benchmarks, see [here](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/benchmark_triton_dropout.py#L90) for instance (it wipes the GPU cache in between iterations to cover that if my memory serves me well)
